### ISSUE TITLE
Support for async driver

### ISSUE THREAD

**Issue by fsbdev** [RobThree/MongoRepository#4]
The underlying driver is at RC for the 2.0 release which uses a fully async stack.
We'll have to update the repository for this release.
https://github.com/mongodb/mongo-csharp-driver/releases

**Comment by RobThree** [comment]
I'm considering a separate branch / project for this; this way we can keep supporting the 1.x driver ("[the 1.x series of the driver will continue to be developed"](https://github.com/mongodb/mongo-csharp-driver/releases)) and the 2.x driver. Any thoughts on this are welcome. PR's too ;-)

**Comment by scratch-net** [comment]
Hi and thanks for not forgetting about us. I also hope to see async 2.0 mongodb driver version support soon.

**Comment by fsbdev** [comment]
The separate branch sounds good.

**Comment by RobThree** [comment]
PR's are very welcome; I am currently very busy and notced that there are pretty much breaking changes...

**Comment by fsbdev** [comment]
The API is very different.. I've started a branch on my fork https://github.com/fsbdev/MongoRepository/tree/v2.x which I've updated for the new API, replacing implementations where I understand what to do in the new API (mostly the collection CRUD calls).
There are many other methods that I'm not sure yet how to implement with the new API, some related to indexes. For those I have simple added `throw new NotImplementedException()` in order to get things to compile in the meantime. I also removed the .NET 3.5 and 4.0 projects from the solution in this branch since the 2.0 driver dropped support for them. If anyone wants to add to my progress please feel welcome. This is just a first dive, I haven't performed any tests on the current change yet.

**Comment by fsbdev** [comment]
I've added support for EnsureIndex now (which uses the CreateIndex method underneath). https://github.com/fsbdev/MongoRepository
Still quite a bit to go but could be enough to start testing for some. Contributions welcome.

**Comment by fsbdev** [comment]
I should point out though that the 2.0 driver doesn't support LINQ as 1.0 did, they are rewriting the LINQ provider and will have support back in 2.1.

**Comment by RobThree** [comment]
> I should point out though that the 2.0 driver doesn't support LINQ as 1.0 did, they are rewriting the LINQ provider and will have support back in 2.1.

Yeah, I found that out too. Maybe it'd be wise to wait for 2.1. Not sure.

**Comment by fsbdev** [comment]
Yeah I probably will.

**Comment by amatlock930** [comment]
Is there any progress on supporting the latest mongo driver and async operations? I see there are some forks that seem to be working on this, but they don't seem to be active.

**Comment by fsbdev** [comment]
I was waiting for LINQ support to come back (2.1 last time I checked).

**Comment by cccsdh** [comment]
https://github.com/cccsdh/NLog.MongoDB20 - you could use this fork of the project...

**Comment by vegar** [comment]
Wouldn't it at least be an option to compile MongoRepository against the legacy driver that comes with the 2.0.* packages? That way it's possible to have MongoRepository co-exists with code that uses the new driver.

**Comment by RobThree** [comment]
@vegar You could try to update the driver and run the unittests and let me know; I am currently very busy selling my house, moving and working so it may be a little while before I get around to it. Sorry.

**Comment by vegar** [comment]
I have a local branch that runs the tests ok, but I had to remove net3.5/4.0 support.

**Comment by RobThree** [comment]
I could be mistaken but some methods (like `GetTotalDataSize()`, `GetTotalStorageSize()`, `RequestStart()()` and `RequestDone()`) aren't supported anymore under 2.0.1 and `GetServer()` is obsoleted, isn't it?
We could use the 1.9.X for 3.5/4.0 support.

**Comment by vegar** [comment]
Yes, the methods mentioned are gone, but I didn't see any problems with GetServer().

I have made a pull request (#11) with my changes, so if you have some time in beween selling houses and getting some work done... ;-)

**Comment by RobThree** [comment]
I appreciate the PR but I do have a few issues with it. Sorry.

* There's no need to remove the `GetTotalDataSize()` and `GetTotalStorageSize()`; these are still available but via `GetStats().DataSize` and `GetStats().StorageSize`.
* I'm not sure support for `RequestStart()` and `RequestDone()` is still there via alternative methods or if it has been dropped; in the latter case we should (I think) also drop the `RequestStart()` and `RequestDone()` methods. However, if there's still support via alternative methods we should at least try to make an effort to implement it without changing/breaking MongoRepository's public API.
* I think simply dropping support for .Net 3.5/4.0 is, also, something that shouldn't be taken lightly; if there's a way to still support 3.5/4.0 we should make an effort. [As I mentioned](https://github.com/RobThree/MongoRepository/issues/4#issuecomment-135426017), we may still be able to provide support by using the 1.9.x driver for those platforms.

Not that I'm complaining, I'm grateful for the effort and PR, but I'd rather take a little more time than breaking people's projects when there's (possible) alternatives.

**Comment by CKoewing** [comment]
Hi Rob,

**FYI**:
> Robert Stam (from MongoDB):
> > Yes, LINQ support is planned for version 2.1.
> 
> Me:
> > When will Version 2.1 come out?
>
> Andy Kor (from MongoDB):
> > No date, but....
> > 
> > We are almost feature complete for 2.1. LINQ is in Code Review right now, GridFS is getting finished up. We'll push a release candidate first to get feedback. 


Best regards,
Chris

**Comment by RobThree** [comment]
Sweet!

**Comment by chemitaxis** [comment]
Hi guys, I was using MongoRepository 1.6.4.0 and all was working fine. My Database Provider has migrate the databases to version 3.0, and all my querys is not working now. I have updated the driver to connect to 3.0, and MongoRepository is not working now (bad results)... I have read all about this, but I don't know if I can test or something to make these MongoRepositry works fine... Thanks!!!!

**Comment by RobThree** [comment]
You could try 1.6.8 (maybe force driver version 1.10 using `Install-Package mongocsharpdriver -Version 1.10.1`). Other than that I'm afraid we're going to sit this one out; we rely too much on the underlying driver's functionality to be able to provide all features we currently do. You could also define "bad results"; I'm not sure how to interpret that. Do you get errors? Unexpected results (but results nonetheless) or...?

**Comment by vegar** [comment]
I guess the most important thing to clarify is if the csharp-driver/mongorepository creates a different query now then before, or if mongo 3.0 executes the query differently. 

You can turn on full profiling on the mongo server to see what query is executed.

Upgrading the server to version 3.0 should not affect anything, though. In theory...

**Comment by CKoewing** [comment]
[MongoDB .NET Driver Version 2.1.0-rc0](https://github.com/mongodb/mongo-csharp-driver/releases/tag/v2.1.0-rc0) is out. Seen? Go, go, go! :-P

**Comment by mellodev** [comment]
2.2.0-rc0 was released yesterday, which introduces sync versions of all async methods:

"The 2.0 and 2.1 versions of the .NET driver featured a new async-only API. Some users gave us feedback that they wanted a choice whether to use a sync or an async API. Version 2.2 introduces sync versions of every async method."

https://github.com/mongodb/mongo-csharp-driver/releases/tag/v2.2.0-rc0

**Comment by chemitaxis** [comment]
Hi guys!! What is the real state of this project? Thanks!!

**Comment by RobThree** [comment]
PR's are welcome; other than that the [current branches](https://github.com/RobThree/MongoRepository/network) show the state as it currently is.

I still aim to get this done rather sooner than later but I've been very busy and with Christmas etc. coming up I don't see much improvement for the short term. Sorry; but again: feel free to start work and submit a PR :wink:

**Comment by RobThree** [comment]
I've started a [V2 branch](https://github.com/RobThree/MongoRepository/tree/v2) and started initial work on it. But I'm still getting used to the V2 C# driver and haven't implemented everything yet. I hope to get some work done during the holidays but don't hold your breath... :stuck_out_tongue:

**Comment by RobThree** [comment]
I think I made some [nice progress](https://github.com/RobThree/MongoRepository/commit/c901f60600aae45db34c9aab74d7f09e9ae06103). The sync part of the Repository should be almost done; that would probably be a 2.0 release and for 2.1 we could focus on Async support and maybe optimizing / cleaning up the current code a bit.

**Comment by chemitaxis** [comment]
Thank you very much!!! ;) very good job I will try to help you with some
tasks!!

**Comment by RobThree** [comment]
I think I have implemented the 2.0 version for the most part (the sync part that is). But I haven't tested anything yet and the (poor) current unittests (although all green) really deserve to be dropped entirely and rewritten to a decent test-suite.

I kept 2.0 (let's call it 'alpha' for the time being) mostly backward compatible. Most notable changes are:

**MongoRepository**
* Dropped: `RequestStart()` and `RequestDone()` methods
* `Collection` property *was* of type `MongoCollection<T>`, now `IMongoCollection<T>`

**MongoRepositoryManager**
* Dropped: `EnsureIndexes(IMongoIndexKeys keys, IMongoIndexOptions options)` overload
* `GetIndexes()` method *was* of type `GetIndexesResult`, now `List<BsonDocument>`

Particularly the `GetIndexes()` method I would like to have another look at to maybe wrap it in a `GetIndexesResult` class from the MongoRepository namespace, as I've done with `CollectionStatsResult` and `ValidateCollectionResult` classes (which, by the way, are now a copy from the [mongo-csharp-driver](https://github.com/mongodb/mongo-csharp-driver) and will probably need to be re-implemented/re-written for licensing issues..?).

So there's still work left to be done; in order of importance:

* Unittests, testing, testing, testing!
* Try to 'fix' the `GetIndexes()` method on the `MongoRepositoryManager` to again return a `GetIndexesResult` for (better) backward compatibility
* Go over code to look for improvements (particularly in the performance department)
* Async support (2.1)

I have rewritten the V2 branch pretty much from the wrist without deep knowledge of the new driver and am quite sure there's lots of room for improvement. Also, I havent tested many things (yet), particularly in the `MongoRepositoryManager` department.

Anyone and everyone who wants to give the V2 branch a whirl to see how it performs, what works, what is broken etc. is very welcome! Simply clone, compile and reference the V2 in your (toy)project.

**We are still very much in alpha, DO NOT use in production or for anything serious other than for testing purposes!**

**Comment by RobThree** [comment]
Aside from PR #8 I'm curious if there's any feedback on the progress so far. I'd like to begin finalizing and work towards a 2.0 release.

**Comment by ukurt** [comment]
hello,

Is there any progress about IMongoCollection version ? We will go to production but we are waiting for your new API :)

**Comment by RobThree** [comment]
I think I was [pretty clear here](https://github.com/RobThree/MongoRepository/issues/4#issuecomment-188258780) and [here](https://github.com/RobThree/MongoRepository/issues/4#issuecomment-165779769), as is the `help wanted` tag. So maybe you shouldn't wait but help out :wink:

**Comment by nathanyeung** [comment]
@RobThree does MongoRepository have LINQ support?

**Comment by RobThree** [comment]
> @RobThree does MongoRepository have LINQ support?

It does. Why would you think it didn't? Did you even try it out?

**Comment by nathanyeung** [comment]
Yes I did and got an exception. Thanks 

Nathan Yeung




On Fri, Aug 12, 2016 at 4:45 PM -0700, "Rob Janssen" <notifications@github.com> wrote:













@RobThree does MongoRepository have LINQ support?




It does. Why would you think it didn't? Did you even try it out?

**Comment by RobThree** [comment]
"An exception"
Without code and the actual exception I won't be able to help. Please open a new issue with the required information/code to reproduce and the actual exception message. You're not supposed to "hijack" some random other issue for help on your problem. This issue is about async support for the 2.0 driver.

**Comment by nathanyeung** [comment]
Thanks for the reply sorry for hijacking an issue. Take care

Nathan Yeung




On Fri, Aug 12, 2016 at 6:12 PM -0700, "Rob Janssen" <notifications@github.com> wrote:












"An exception"

Without code and the actual exception I won't be able to help. Please open a new issue with the required information/code to reproduce and the actual exception message. You're not supposed to "hijack" some random other issue for help on your problem. This issue is about async support for the 2.0 driver.

**Comment by ducas** [comment]
hi rob. are you just waiting on someone to validate this? do you have a nuget package around for the v2 branch? if not can you put up a prerelease build somewhere so we can test it easily?
